emerge-gitclone

```diff
@@ -41,7 +41,8 @@ for repo in portage.db[eroot]['vartree'].settings.repositories:
 	if commit:
 		commit = commit.replace('refs/heads/', 'refs/remotes/origin/')
 	else:
-		print("Warning: No revision found for " + repo.sync_uri)
+		print("Warning: No revision found for %s. Falling back to flatcar-master." % repo.sync_uri)
+		commit = "flatcar-master"
```

Comment:
Which package was this about? `flatcar-master` could also be wrong and I wonder if we should rather throw an error here instead of continuing to build something that was not planned to be built.

Reply:
Right, I fixed the branch name to `main`.

We should continue building in that case.
`commit` being invalid does not mean an error case, but every non-release build, e.g. nightly CI builds.
In case of release builds, `release` == `3127.0.0`, so it is possible to get manifest from https://raw.githubusercontent.com/kinvolk/manifest/v3127.0.0/release.xml. As a result, `commit` becomes valid.
In case of non-release builds, `release` is not a semver like that, but a dev build number like `2022.02.02+dev-flatcar-master-4742`. So it is not possible to get release.xml. As a result, `commit` becomes invalid.

You will see strings like `Failed to read manifest commits for 2022.02.02+dev-flatcar-master-4742`, e.g. in http://jenkins.infra.kinvolk.io:8080/job/os/job/kola/job/dev-container/3372/.

Reply:
Ok, got it, then the solution I suggest here is to have an `else:` branch for `if release:` (see the block above)

Reply:
Also, we should plan to migrate this to the new submodule approach the SDK container uses (no manifests anymore).

Reply:
Do you mean [this block](https://github.com/flatcar-linux/flatcar-dev-util/blob/74d8f0b1814b5c205a7340d1f4204f9082a9eda7/emerge-gitclone#L19)?

I am not sure what we can do in that phase.
`commits` cannot be accessed, as it is not possible to get `repo.getAttribute('name')`.
No release.xml, no repo, no commits.
So later `commit` becomes always invalid in [every run](https://github.com/flatcar-linux/flatcar-dev-util/blob/74d8f0b1814b5c205a7340d1f4204f9082a9eda7/emerge-gitclone#L34).

Reply:
`main` is not the expected branch, we can get the right one from the dev build manifest. The url should not be `https://raw.githubusercontent.com/kinvolk/manifest-builds/v%s/release.xml` but `https://raw.githubusercontent.com/kinvolk/manifest-builds/%s/release.xml` with `%s` being the `dev-flatcar-master-4742` suffix of the `2022.02.02+dev-flatcar-master-4742` version

Reply:
```
diff --git a/emerge-gitclone b/emerge-gitclone
index f0731c0..e034e19 100755
--- a/emerge-gitclone
+++ b/emerge-gitclone
@@ -17,7 +17,10 @@ with open('/usr/share/flatcar/release') as release_file:
 # Attempt to read Git commits from this release's manifest.
 commits = {}
 if release:
-       manifest = "https://raw.githubusercontent.com/kinvolk/manifest/v%s/release.xml" % release
+       if "+" in release:
+               manifest = "https://raw.githubusercontent.com/kinvolk/manifest-builds/%s/release.xml" % release.split("+")[1]
+       else:
+               manifest = "https://raw.githubusercontent.com/kinvolk/manifest/v%s/release.xml" % release
        try:
                from xml.dom.minidom import parseString as pxs
                from urllib import request as req
@@ -25,6 +28,7 @@ if release:
                        commits[repo.getAttribute('name')] = repo.getAttribute('revision')
        except Exception:
                print(">>> Failed to read manifest commits for %s" % release)
+               sys.exit(1)
 
 synced = False
 eroot = portage.settings['EROOT']
```

Reply:
Maybe we can leave out the `sys.exit(1)` but I'm in favor of erroring out early instead of having misbehavior like this lingering around until it breaks like now

Reply:
(I expect using `main` to fail for those repos where we only have `flatcar-master`, anyway we should not work around the underlying problem, but need to find the right way of checking out what should be checked out)

Reply:
That sounds like a good idea.
Though I had to change the URL to https://raw.githubusercontent.com/kinvolk/manifest-builds/dev-flatcar-master-4742/dev-flatcar-master-4742.xml, as https://raw.githubusercontent.com/kinvolk/manifest-builds/dev-flatcar-master-4742/release.xml is an invalid URL.

```
+       if "+" in release:
+               dev_build_tag = release.split("+")[1]
+               manifest = "https://raw.githubusercontent.com/kinvolk/manifest-builds/%s/%s.xml" % (dev_build_tag, dev_build_tag)
+       else:
+               manifest = "https://raw.githubusercontent.com/kinvolk/manifest/v%s/release.xml" % release
```

> I expect using main to fail for those repos where we only have flatcar-master

Yeah, in theory you are right.
But the simple `main` had worked as well, because the script is cloning only 2 git repos, coreos-overlay and portage-stable.

Reply:
Good, thanks!